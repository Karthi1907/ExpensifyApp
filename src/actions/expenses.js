import uuid from 'uuid';

export const addExpense = 
    ( 
        {
            descr, 
            comments,
            amount = 0,
            createdAt = 0
        } = {} 
    ) =>
({
    type: 'ADD_EXPENSE',
    expense:
    {
        id: uuid(),
        descr,
        comments,
        amount,
        createdAt
    }
});

export const removeExpense = ( id ) => (
    {
        type: 'DELETE_EXPENSE',
        id
    }
)

export const editExpense = ( (id, updates ) => 
    (
        {
            type: 'EDIT_EXPENSE',
            id,
            updates
        }
    )
);
