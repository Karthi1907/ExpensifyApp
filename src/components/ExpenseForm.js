import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


const now = moment();
console.log(now.format('YYYY MMMM Do'));   
console.log(now.startOf('month').format('YYYYMMDD'));


export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            descr: props.expense ? props.expense.descr : '',
            comments: props.expense ? props.expense.comments : '',
            amount: props.expense ? props.expense.amount : '0.00',
            createdAt: props.expense ? moment(props.expense.createdAt, 'YYYYMMDD')   : moment(),
            calendarFocused: false,
            error: undefined
        }
    }



    onDescriptionChange = (e) => {

        const description = e.target.value;
        this.setState( () => ({ descr: description }) );
        console.log(this.state);
    };

    onCommentsChange = ( e ) => {
        //const comments = e.target.value;
        e.persist();
        this.setState( () => ( { comments: e.target.value }) );
        console.log(this.state);
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount ||  amount.match(/^\d{1,7}(\.\d{0,2})?$/) ) {
            this.setState( () => ({  amount }) );
        }
    }

    onDateChange = ( createdAt ) => {
        
        if (createdAt) {
            const createdAtDate = createdAt.format('YYYYMMDD');
            console.log(createdAtDate);
            this.setState( () => ( { createdAt } ) );
            console.log(createdAt.format('YYYYMMDD'));
            console.log(this.state);    
            console.log(this.state.createdAt);
        }
    }

    onFocusChange = ( {focused} ) => {
        this.setState( () => ( { calendarFocused: focused } ) );
        console.log(this.state);
    }

    onSubmit = (e) => {

        e.preventDefault();
        console.log('Inside onSubmit() function');
        let error = '';

        if ( !this.state.descr || !this.state.amount ) {
            error = 'Mandatory inputs missing - add both description and amount to continue';
        }

        if (error) {
            this.setState( () => ({ error: error }) );
        } else {
            this.setState( () => ({ error: '' }) );
        
            console.log(error);
            console.log(this.state); 
            this.props.onSubmit( {
                descr: this.state.descr,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.format('YYYYMMDD'),
                comments: this.state.comments
            } ) ;  
        }
    };


    render()
    {
        return (
            <form onSubmit={this.onSubmit}> 
            { this.state.error && <p style={{color: 'red',fontWeight:'bold'}}>  {this.state.error} </p> }   
            <div>
                <div>
                    <p> Description: </p>
                    <input 
                        type="text"
                        placeholder="Your Expense Descr"
                        autoFocus
                        value={this.state.descr}
                        onChange={this.onDescriptionChange}
                    />
                </div>
                <div>
                    <p> Amount:</p>
                    <input 
                        type="text"
                        value={this.state.amount}
                        placeholder="Your Expense Amount"
                        onChange={this.onAmountChange}
                    />
                </div>
                <div>
                <p> Create Date: </p>
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    displayFormat="DD-MM-YYYY"
                    numberOfMonths={1}
                    isOutsideRange={ () => false }
                />
                </div>
                <div>
                    <p> Comments</p>
                    <textarea
                        type="text"
                        placeholder="Add Comments for your expense here"
                        value={this.state.comments}
                        onChange={this.onCommentsChange}
                    />
                </div>
                <div> 
                    <p> </p>
                    <button> Add Expense </button>
                </div> 
            </div>
            </form>
        );
    }
};