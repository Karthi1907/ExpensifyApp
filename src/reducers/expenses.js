const expenseDefaultState = [];

export default (expenses = expenseDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            console.log(`ADDNIG EXPENSE  ${action.expense.descr}`)
            //return expenses.concat(action.expense) -- this is okay as well but we will use spread operator to get same result as below
            return [...expenses, action.expense]

        case 'DELETE_EXPENSE':
            console.log(`DELETING EXPENSE ${action.id}`) 
            // return expenses.filter( (expense) => {
            //         return expense.id !== action.id;
            //     }
            // );   
            return expenses.filter( ( {id} )  => id !== action.id);

        case 'EDIT_EXPENSE':
            console.log(`Editing Expense id: ${action.id} `)
            return expenses.map( ( expense ) => {
                if ( expense.id === action.id ) 
                {
                    return { ...expense, ...action.updates }
                } 
                else 
                {
                    return expense;
                }  
            });
        
        default:
            return expenses;
    }
};