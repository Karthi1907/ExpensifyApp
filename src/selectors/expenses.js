// Get Visible Expenses

export default ( ( expenses, {text, sortBy, startDate, order, endDate} ) => 
    {
        return expenses.filter(( expense ) => {
            const startDateFormatted = startDate ? startDate.format('YYYYMMDD') : null;
            const endDateFormatted = endDate ? endDate.format('YYYYMMDD') : null;
            console.log (` startDate: ${startDateFormatted} | endDate: ${endDateFormatted} | text: ${text} | sortBy: ${sortBy}
               order: ${order} | expense descr: ${expense.descr} | expense createdAt: ${expense.createdAt} `);
            const startDateMatch = startDate ? expense.createdAt >= startDateFormatted : true;
            const endDateMatch = endDate ? expense.createdAt < endDateFormatted : true;
            const textMatch =  !text ||  expense.descr.toLowerCase().includes(text.toLowerCase()) ;
            console.log (`${startDateMatch} && ${endDateMatch} && ${textMatch}`)

            return startDateMatch && endDateMatch && textMatch;
        }).sort( (a,b) => {

                switch(sortBy) {
                    case 'date':
                        switch (order) {
                            case 'asc':
                                return (
                                    a.createdAt === b.createdAt ? 0 : a.createdAt  > b.createdAt ? 1 : -1
                                );
                            case 'desc':
                                return (
                                    a.createdAt === b.createdAt ? 0 : a.createdAt < b.createdAt ? 1 : -1        
                                );
                        }
                    case 'amount':
                    default:
                        switch (order) {
                            case 'asc':    
                                return (
                                    a.amount === b.amount ? 0 : a.amount > b.amount ? 1 : -1
                                );
                            case 'desc':
                                return (
                                    a.amount === b.amount ? 0 : a.amount < b.amount ? 1 : -1
                                );
                        }
                } 

            } 
        )
    }   
);