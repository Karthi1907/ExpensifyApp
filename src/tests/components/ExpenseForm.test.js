import React from 'react';
import ExpenseForm from '../../components/ExpenseForm';
import { shallow } from 'enzyme';
import { defaultTestExpenses } from '../fixtures/expenses';
import moment from 'moment';
const expenseOne = { id: 66,  amount: 100, descr: 'EBB Bill', createdAt: '20230701', comments: 'Nothing Nothing' };


test('ExpenseForm should be rendered correctly', () => {
    const wrapper = shallow( <ExpenseForm />);
    expect(wrapper).toMatchSnapshot();
});

test('ExpenseForm should be rendered correctly - TC 2', () => {
   
    const wrapper = shallow( <ExpenseForm expense={expenseOne}/>);
    expect(wrapper).toMatchSnapshot();
}); 

test('ExpenseForm should be rendered correctly - TC 3 - Form Submit - No decsr/amount', () => {
    const wrapper = shallow( <ExpenseForm />);
    // expect(wrapper).toMatchSnapshot);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
}); 

test('should set descr(iption) on input text change', () => {
    const value = 'Test Description';
    const wrapper = shallow( <ExpenseForm />);
    // console.log(wrapper.find('input').at(0));
    // expect(wrapper).toMatchSnapshot);
    wrapper.find('input').at(0).simulate('change', {
        target : {value}
    });
    expect(wrapper.state('descr')).toBe(value);
    // expect(wrapper).toMatchSnapshot();
}); 

test('should set comments on input textarea change', () => {
    const value = 'Test Comments';
    const wrapper = shallow( <ExpenseForm />);
    // expect(wrapper).toMatchSnapshot);
    wrapper.find('textarea').simulate('change', {
        target : {value},
        persist: jest.fn()
    });
    expect(wrapper.state('comments')).toBe(value);
}); 

test('should set amount only if valid amount', () => {
    const value = '123.456';
    const wrapper = shallow( <ExpenseForm />);
    // expect(wrapper).toMatchSnapshot);
    wrapper.find('input').at(1).simulate('change', {
        target : {value}
    });
    expect(wrapper.state('amount')).toBe('0.00');
}); 

test('should set amount only if valid amount', () => {
    const value = '123.45';
    const wrapper = shallow( <ExpenseForm />);
    // expect(wrapper).toMatchSnapshot);
    wrapper.find('input').at(1).simulate('change', {
        target : {value}
    });
    expect(wrapper.state('amount')).toBe(value);
}); 

test(' should call submit prop for valid form submission ', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<ExpenseForm expense={expenseOne} onSubmit={onSubmitSpy} />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        amount: 100, descr: 'EBB Bill', createdAt: '20230701', comments: 'Nothing Nothing'
    });
});

test(' should check SingleDatePicker onDateChange is successful ', () => {
  console.log('should check SingleDatePicker onDateChange is successful');
  const now = moment();
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('SingleDatePicker').prop('onDateChange')(now);
  expect(wrapper.state('createdAt')).toEqual(now);  
});

test(' should check SingleDatePicker onFocusChange is successful ', () => {
    const calendarFocused = false;
    console.log('should check SingleDatePicker onFocusChange is successful');
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('SingleDatePicker').prop('onFocusChange')({focused: calendarFocused});
    expect(wrapper.state('calendarFocused')).toBe(calendarFocused);  
  });