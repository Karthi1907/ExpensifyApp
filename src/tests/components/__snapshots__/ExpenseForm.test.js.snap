// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseForm should be rendered correctly - TC 2 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <div>
      <p>
         Description: 
      </p>
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Your Expense Descr"
        type="text"
        value="EBB Bill"
      />
    </div>
    <div>
      <p>
         Amount:
      </p>
      <input
        onChange={[Function]}
        placeholder="Your Expense Amount"
        type="text"
        value={100}
      />
    </div>
    <div>
      <p>
         Create Date: 
      </p>
      <SingleDatePicker
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"2023-06-30T18:30:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat="DD-MM-YYYY"
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div>
      <p>
         Comments
      </p>
      <textarea
        onChange={[Function]}
        placeholder="Add Comments for your expense here"
        type="text"
        value="Nothing Nothing"
      />
    </div>
    <div>
      <p>
         
      </p>
      <button>
         Add Expense 
      </button>
    </div>
  </div>
</form>
`;

exports[`ExpenseForm should be rendered correctly - TC 3 - Form Submit - No decsr/amount 1`] = `
<form
  onSubmit={[Function]}
>
  <p
    style={
      Object {
        "color": "red",
        "fontWeight": "bold",
      }
    }
  >
      
    Mandatory inputs missing - add both description and amount to continue
     
  </p>
  <div>
    <div>
      <p>
         Description: 
      </p>
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Your Expense Descr"
        type="text"
        value=""
      />
    </div>
    <div>
      <p>
         Amount:
      </p>
      <input
        onChange={[Function]}
        placeholder="Your Expense Amount"
        type="text"
        value="0.00"
      />
    </div>
    <div>
      <p>
         Create Date: 
      </p>
      <SingleDatePicker
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"2024-01-16T07:25:20.441Z"}
        daySize={39}
        disabled={false}
        displayFormat="DD-MM-YYYY"
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div>
      <p>
         Comments
      </p>
      <textarea
        onChange={[Function]}
        placeholder="Add Comments for your expense here"
        type="text"
        value=""
      />
    </div>
    <div>
      <p>
         
      </p>
      <button>
         Add Expense 
      </button>
    </div>
  </div>
</form>
`;

exports[`ExpenseForm should be rendered correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <div>
      <p>
         Description: 
      </p>
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Your Expense Descr"
        type="text"
        value=""
      />
    </div>
    <div>
      <p>
         Amount:
      </p>
      <input
        onChange={[Function]}
        placeholder="Your Expense Amount"
        type="text"
        value="0.00"
      />
    </div>
    <div>
      <p>
         Create Date: 
      </p>
      <SingleDatePicker
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"2024-01-16T07:25:20.436Z"}
        daySize={39}
        disabled={false}
        displayFormat="DD-MM-YYYY"
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div>
      <p>
         Comments
      </p>
      <textarea
        onChange={[Function]}
        placeholder="Add Comments for your expense here"
        type="text"
        value=""
      />
    </div>
    <div>
      <p>
         
      </p>
      <button>
         Add Expense 
      </button>
    </div>
  </div>
</form>
`;
